src/features/device-state/routes/state-patch-v3.ts

```diff
@@ -0,0 +1,384 @@
+import type { RequestHandler } from 'express';
+
+import * as _ from 'lodash';
+import {
+	captureException,
+	handleHttpErrors,
+} from '../../../infra/error-handling';
+import { sbvrUtils, errors } from '@balena/pinejs';
+import { getIP } from '../../../lib/utils';
+import {
+	Application,
+	Device,
+	Image,
+	ImageInstall,
+	PickDeferred,
+	Release,
+} from '../../../balena-model';
+import { deleteOldImageInstalls, upsertImageInstall } from './state-patch-v2';
+import type { Filter } from 'pinejs-client-core';
+import { metricsPatchFields, v3ValidPatchFields } from '..';
+import { shouldUpdateMetrics } from '../utils';
+
+const { BadRequestError, UnauthorizedError, InternalRequestError } = errors;
+const { api } = sbvrUtils;
+
+/**
+ * These typings should be used as a guide to what should be sent, but cannot be trusted as what actually *is* sent.
+ */
+export type StatePatchV3Body = {
+	[uuid: string]: {
+		name?: string;
+		status?: string;
+		os_version?: string;
+		os_variant?: string;
+		supervisor_version?: string;
+		provisioning_progress?: number | null;
+		provisioning_state?: string | null;
+		ip_address?: string;
+		mac_address?: string;
+		api_port?: number;
+		api_secret?: string;
+		logs_channel?: string; // TODO: should this not be reported anymore?
+		memory_usage?: number;
+		memory_total?: number;
+		storage_block_device?: string;
+		storage_usage?: number;
+		storage_total?: number;
+		cpu_temp?: number;
+		cpu_usage?: number;
+		cpu_id?: string;
+		is_undervolted?: boolean;
+		/**
+		 * Used for setting dependent devices as online
+		 */
+		is_online?: boolean;
+		/**
+		 * Used for setting gateway device of dependent devices
+		 */
+		is_managed_by__device?: number;
+		apps?: {
+			[uuid: string]: {
+				/**
+				 * We report the overall release uuid the supervisor considers the active one, even though there may be info for multiple releases.
+				 */
+				release_uuid?: string;
```

Comment:
```suggestion
				current_release_uuid?: string;
```

?

Or:

```suggestion
				active_release_uuid?: string;
```

?

Reply:
The idea is for the current and target states to use a matching format and so things like `current`/`active` don't make sense, there's just a state that happens to be what the device is currently running and one that is what it should be running, and the supervisor just does what is necessary to transform one to the other. In practice there are minor differences where eg the target state doesn't specify things like `is_undervolted` because it doesn't really make sense to do so (it's not like the supervisor can control it), and the current state doesn't report back things like env vars/config because the api already knows it and it'd be a waste of bandwidth (but again it's just an optimization and it could report it and the api would just ignore it)